ouroboros-network-framework/src/Ouroboros/Network/Socket.hs

```diff
@@ -241,13 +247,17 @@ connectToNode' sn versionDataCodec NetworkConnectTracers {nctMuxTracer, nctHands
     ts_end <- getMonotonicTime
 
     case mapp of
-         Left (errDrv :: ProtocolLimitFailure)-> do
+         Nothing -> do
+            traceWith muxTracer $ Mx.MuxTraceHandshakeClientError ExceededTimeLimit
+              (diffTime ts_end ts_start)
+            throwIO ExceededTimeLimit
+         Just (Left (errDrv :: ProtocolLimitFailure))-> do
             traceWith muxTracer $ Mx.MuxTraceHandshakeClientError errDrv (diffTime ts_end ts_start)
             throwIO errDrv
-         Right (Left err) -> do
+         Just (Right (Left err)) -> do
             traceWith muxTracer $ Mx.MuxTraceHandshakeClientError err (diffTime ts_end ts_start)
             throwIO err
-         Right (Right app) -> do
+         Just (Right (Right app)) -> do
```

Comment:
Can we simplify the `Maybe (Either _ (Either _ r))` type?

Reply:
what about this idea:

```
data HandshakeException a = 
    HandshakeProtocolLimit ProtocolLimitfailure
  | HandshakeProtocolError a
  | HanshakeTimeout
 
tryHandshake ::  m (Either a r) -> m (Either (HanshakeException a) r)
tryHandshaket = ... withTimeoutSerial ... try ... 
```

The `a` in `tryHandshake` would be different for `client` and `server`.

Reply:
A great idea!

Reply:
Really? Even though `HandshakeException a` is polymorphic?

Reply:
I tried and failed to get around the need for a `Maybe (Either _ (Either _ r))` case statement in `tryHandshake` which made things far worse.
The updated code works and is more like your initial suggestion.